src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void StartRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
+            CurlResponseMessage responseMessage = null;
             SafeCurlHandle requestHandle = null;
+            GCHandle requestStateHandle = new GCHandle();
+            IntPtr headerPtr = IntPtr.Zero;
+            bool referenceTaken = false;
+
+#if DEBUG
+            if (Logging.On) Logging.Enter(Logging.Http, this, "StartRequest", state.RequestMessage);
+#endif
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+            var cancellationTokenRegistration = state.CancellationToken.Register(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.TrySetCanceled(rcs.CancellationToken);
```

Comment:
I noted this in a previous PR, but this looks very dangerous.  If a cancellation request arrives, we're going to immediately complete the returned Task as canceled, even if work is still in flight, which means the caller awaiting the operation could then proceed thinking all work is done even though that's not true.  If this is just here temporarily until the dev work is actually done to cancel the operation (at which point the task should only transition to Canceled once all work has quiesced), then there should at least be a TODO comment here so that we don't lose track of it.